<launch>
  <!--
       Required arguments:
          HAND:=right|left
       Suggested arguments:
          ZEROS:=/path/to/zeros_file.yaml
          CONTROLLER:=grasp|pd|velsat
          RESPAWN:=true|false   Respawn controller if it dies.
          KEYBOARD:=true|false  (default is true)
          CAN_DEVICE:=/dev/pcanusb1 | /dev/pcanusbNNN  (ls -l /dev/pcan* to see open CAN devices)
          VISUALIZE:=true|false  (Launch rviz)

       This script launches the following nodes:
         - allegro hand controller (different controllers exist)
         - keyboard controller
         - state publisher (for TF information)

  -->

  <!-- Visualization with rviz, only if arg VISUALIZE is set to true. Default is
       false, the allegro_viz.launch can be started separated. -->
  <arg name ="VISUALIZE" default="false" />
  <include file="$(find allegro_hand_controllers)/launch/allegro_viz.launch"
           if="$(arg VISUALIZE)"/>

  <arg name="CONTROLLER" default="grasp"/> <!-- grasp, pd, velsat  -->
  <!-- Controllers include:
       -grasp *
       -pd
       -velsat
       *The default controller is 'grasp' which employs the included grasping library (libBHand). -->

  <arg name="POLLING" default="true"/> <!-- true, false for polling the CAN communication -->

  <!-- The inclusion of which_hand in the zero.yaml file has been deprecated.
       Which hand (left/right) must now be specified as an argument when launching the Allegro Hand
       as an input for both the robot_description and for the grasping library controllers. -->

  <arg name="HAND"/>
  <arg name="NUM" default='0'/>

  <!-- ls -l /dev/pcan* to see your open CAN ports -->
  <arg name="CAN_DEVICE" default="/dev/pcanusb1"/>

  <arg name="PARAMS_DIR" default="$(find allegro_hand_parameters)" />
  <arg name="KEYBOARD" default="true" />

  <!-- yaml param files for your hand can be found in parameters/zero_files/ -->
  <arg name="ZEROS" default="$(arg PARAMS_DIR)/zero.yaml"/>

  <param name="/robot_description"
         textfile="$(find allegro_hand_description)/allegro_hand_description_$(arg HAND).urdf"/>
  <param name="/allegroHand_$(arg NUM)/robot_description"
         textfile="$(find allegro_hand_description)/allegro_hand_description_$(arg HAND).urdf"/>

  <arg name="RESPAWN" default="false"/> <!--If true, respawn the controller if it dies. -->

  <!-- Allegro Hand controller and communication node -->
  <node name="allegroHand_$(arg HAND)_$(arg NUM)"
        pkg="allegro_hand_controllers"
        type="allegro_node_$(arg CONTROLLER)"
        output="screen"
        clear_params="true"
        respawn="$(arg RESPAWN)"
        args="$(arg POLLING)" >

    <!-- Remapping of topics into enumerated allegroHand_# namespace -->
    <remap from="/allegroHand/joint_states" to="/allegroHand_$(arg NUM)/joint_states"/>
    <remap from="/allegroHand/joint_cmd" to="/allegroHand_$(arg NUM)/joint_cmd"/>
    <remap from="/allegroHand/lib_cmd" to="/allegroHand_$(arg NUM)/lib_cmd"/>

    <!--parameters are within the scope of the hand node so that multiple hands can be run at the same time -->
    <rosparam file="$(arg ZEROS)" command="load" />

    <rosparam file="$(arg PARAMS_DIR)/gains_pd.yaml" command="load" />
    <rosparam file="$(arg PARAMS_DIR)/gains_velSat.yaml" command="load" />
    <rosparam file="$(arg PARAMS_DIR)/initial_position.yaml" command="load" />

    <param name="/comm/CAN_CH" value="$(arg CAN_DEVICE)" />
    <param name="/hand_info/which_hand" value="$(arg HAND)" /> <!-- See HAND arg above -->
  </node>

  <!-- Joint States (angles) to Joint Transforms -->
  <node name="jointState2tf_$(arg NUM)"
        pkg="robot_state_publisher"
        type="state_publisher">
    <remap from="/tf" to="/allegroHand_$(arg NUM)/tf"/>
    <remap from="/joint_states" to="/allegroHand_$(arg NUM)/joint_states"/>
  </node>

  <!-- Keyboard handler (if arg KEYBOARD is true) -->
  <node name="keyboard_$(arg NUM)"
        pkg="allegro_hand_keyboard"
        type="allegro_hand_keyboard_node"
        output="screen"
        if="$(arg KEYBOARD)">
    <remap from="/allegroHand/lib_cmd" to="/allegroHand_$(arg NUM)/lib_cmd"/>
  </node>

</launch>
